File: 15 Boxing/python/boxing.py
#!/usr/bin/env python3
import random

QUESTION_PROMPT = '? '

def play():
    print('BOXING')
    print('CREATIVE COMPUTING   MORRISTOWN, NEW JERSEY')
    print('\n\n')
    print('BOXING OLYMPIC STYLE (3 ROUNDS -- 2 OUT OF 3 WINS)')

    opponent_score = 0
    player_score = 0

    opponent_damage = 0
    player_damage = 0

    opponent_knockedout = False
    player_knockedout = False

    print('WHAT IS YOUR OPPONENT\'S NAME', end=QUESTION_PROMPT)
    opponent_name = input()
    print('WHAT IS YOUR MAN\'S NAME', end=QUESTION_PROMPT)
    player_name = input()

    print('DIFFERENT PUNCHES ARE 1 FULL SWING 2 HOOK 3 UPPERCUT 4 JAB')
    print('WHAT IS YOUR MAN\'S BEST', end=QUESTION_PROMPT)
    player_best = int(input())

    print('WHAT IS HIS VULNERABILITY', end=QUESTION_PROMPT)
    player_weakness = int(input())

    opponent_best = 0
    opponent_weakness = 0 
    while opponent_best == opponent_weakness:
        opponent_best = random.randint(1, 4)
        opponent_weakness = random.randint(1, 4)

    print('{}\'S ADVANTAGE is {} AND VULNERABILITY IS SECRET.'
          .format(opponent_name, opponent_weakness))

    for round in (1, 2, 3):
        print('ROUND {} BEGINS...\n'.format(round))
        if opponent_score >= 2 or player_score >= 2:
            break

        for action in range(7):
            if random.randint(1, 10) > 5:
                # opponent swings
                punch = random.randint(1, 4)

                if punch == player_weakness:
                    player_damage += 2

                if punch == 1:
                    print('{} TAKES A FULL SWING AND'.format(opponent_name), end = ' ')
                    if player_weakness == 1 or random.randint(1, 60) < 30:
                        print('POW!!!! HE HITS HIM RIGHT IN THE FACE!')
                        if player_damage > 35:
                            player_knockedout = True
                            break
                        player_damage += 15
                    else:
                        print('BUT IT\'S BLOCKED!')
                elif punch == 2:
                    print('{} GETS {} IN THE JAW (OUCH!)'.format(opponent_name, player_name), end = ' ')
                    player_damage += 7
                    print('....AND AGAIN')
                    if player_damage > 35:
                        player_knockedout = True
                        break
                    player_damage += 5
                elif punch == 3:
                    print('{} IS ATTACKED BY AN UPPERCUT (OH,OH)'.format(player_name))
                    if player_weakness == 3 or random.randint(1, 200) > 75:
                        print('{} BLOCKS AND HITS {} WITH A HOOK'.format(player_name, opponent_name))
                        opponent_damage += 5
                    else:
                        print('AND {} CONNECTS...'.format(opponent_name))
                        player_damage += 8
                else:
                    print('{} JABS AND'.format(opponent_name), end = ' ')
                    if player_weakness == 4 or random.randint(1, 7) > 4:
                        print('BLOOD SPILLS !!!')
                        player_damage += 3
                    else:
                        print('AND IT\'S BLOCKED (LUCKY BLOCK!)')
            else:
                print('{}\'S PUNCH'.format(player_name), end='? ')
                punch = int(input())

                if punch == opponent_weakness:
                    opponent_damage += 2

                if punch == 1:
                    print('{} SWINGS AND'.format(player_name), end = ' ')
                    if opponent_weakness == 1 or random.randint(1, 30) < 10:
                        print('HE CONNECTS!')
                        if opponent_damage > 35:
                            opponent_knockedout = True
                            break
                        opponent_damage += 15
                    else:
                        print('HE MISSES')
                elif punch == 2:
                    print('{} GIVES THE HOOK...'.format(player_name), end = ' ')
                    if opponent_weakness == 2 or random.randint(1, 2) == 1:
                        print('CONNECTS...')
                        opponent_damage += 7
                    else:
                        print('BUT IT\'S BLOCKED!!!!!!!!!!!!!')
                elif punch == 3:
                    print('{} TRIES AN UPPERCUT'.format(player_name), end = ' ')
                    if opponent_weakness == 3 or random.randint(1, 100) < 51:
                        print('AND HE CONNECTS!')
                        opponent_damage += 4
                    else:
                        print('AND IT\'S BLOCKED (LUCKY BLOCK!)')
                else:
                    print('{} JABS AT {}\'S HEAD'.format(player_name, opponent_name), end = ' ')
                    if opponent_weakness == 4 or random.randint(1, 8) < 4:
                        print('AND HE CONNECTS!')
                        opponent_damage += 3
                    else:
                        print('AND IT\'S BLOCKED (LUCKY BLOCK!)')

        if player_knockedout or opponent_knockedout:
            break
        elif player_damage > opponent_damage:
            print('{} WINS ROUND {}'.format(opponent_name, round))
            opponent_score += 1
        else:
            print('{} WINS ROUND {}'.format(player_name, round))
            player_score += 1

    if player_knockedout:
        print('{} IS KNOCKED COLD AND {} IS THE WINNER AND CHAMP'.format(player_name, opponent_name))
    elif opponent_knockedout:
        print('{} IS KNOCKED COLD AND {} IS THE WINNER AND CHAMP'.format(opponent_name, player_name))
    elif opponent_score > player_score:
        print('{} WINS (NICE GOING), {}'.format(opponent_name, player_name))
    else:
        print('{} AMAZINGLY WINS'.format(player_name))

    print('\n\nAND NOW GOODBYE FROM THE OLYMPIC ARENA.')

if __name__ == '__main__':
    play()

File: 15 Boxing/java/Boxing.java
/**
 * Boxing
 *
 * <p>
 * Based on the Basic game of BatNum here
 * https://github.com/coding-horror/basic-computer-games/tree/main/15%20Boxing
 * <p>
 */
public class Boxing {
    
    private static final Basic.Console console = new Basic.Console();
    
    private GameSession session;

    public void play() {
        showIntro();

        loadPlayers();

        console.print("%s'S ADVANTAGE IS %d AND VULNERABILITY IS SECRET.\n", session.getOpponent().getName(), session.getOpponent().getBestPunch().getCode());


        for (int roundNro = 1; roundNro <= 3; roundNro++) {
            if (session.isOver())
                break;

            session.resetPoints();
            console.print("\nROUND %d BEGINS...%n", roundNro);

            for (int majorPunches = 1; majorPunches <= 7; majorPunches++) {
                long i = Basic.randomOf(10);

                if (i > 5) {
                    boolean stopPunches = opponentPunch();
                    if (stopPunches ) break;
                } else {
                    playerPunch();
                }
            }
            showRoundWinner(roundNro);
        }
        showWinner();
    }

    private boolean opponentPunch() {
        final Punch punch = Punch.random();

        if (punch == session.getOpponent().getBestPunch()) session.addOpponentPoints(2);

        if (punch == Punch.FULL_SWING) {
            console.print("%s TAKES A FULL SWING AND", session.getOpponent().getName());
            long r6 = Basic.randomOf(60);

            if (session.getPlayer().hitVulnerability(Punch.FULL_SWING) || r6 < 30) {
                console.print(" POW!!!!! HE HITS HIM RIGHT IN THE FACE!\n");
                if (session.getPoints(session.getOpponent()) > 35) {

                    session.setKnocked();
                    return true;
                }
                session.addOpponentPoints(15);
            } else {
                console.print(" IT'S BLOCKED!\n");
            }
        }

        if (punch == Punch.HOOK  || punch == Punch.UPPERCUT) {
            if (punch == Punch.HOOK) {
                console.print("%s GETS %s IN THE JAW (OUCH!)\n", session.getOpponent().getName(), session.getPlayer().getName());

                session.addOpponentPoints(7);
                console.print("....AND AGAIN!\n");

                session.addOpponentPoints(5);
                if (session.getPoints(session.getOpponent()) > 35) {
                    session.setKnocked();
                    return true;
                }
                console.print("\n");

            }
            console.print("%s IS ATTACKED BY AN UPPERCUT (OH,OH)...\n", session.getPlayer().getName());
            long q4 = Basic.randomOf(200);
            if (session.getPlayer().hitVulnerability(Punch.UPPERCUT) || q4 <= 75) {
                console.print("AND %s CONNECTS...\n", session.getOpponent().getName());

                session.addOpponentPoints(8);
            } else {
                console.print(" BLOCKS AND HITS %s WITH A HOOK.\n", session.getOpponent().getName());

                session.addPlayerPoints(5);
            }
        }
        else {
            console.print("%s JABS AND ", session.getOpponent().getName());
            long z4 = Basic.randomOf(7);
            if (session.getPlayer().hitVulnerability(Punch.JAB))

                session.addOpponentPoints(5);
            else if (z4 > 4) {
                console.print(" BLOOD SPILLS !!!\n");

                session.addOpponentPoints(5);
            } else {
                console.print("IT'S BLOCKED!\n");
            }
        }
        return true;
    }

    private void playerPunch() {
        console.print("%s'S PUNCH? ", session.getPlayer().getName());
        final Punch punch = Punch.fromCode(console.readInt());

        if (punch == session.getPlayer().getBestPunch()) session.addPlayerPoints(2);

        switch (punch) {
            case FULL_SWING -> {
                console.print("%s SWINGS AND ", session.getPlayer().getName());
                if (session.getOpponent().getBestPunch() == Punch.JAB) {
                    console.print("HE CONNECTS!\n");
                    if (session.getPoints(session.getPlayer()) <= 35) session.addPlayerPoints(15);
                } else {
                    long x3 = Basic.randomOf(30);
                    if (x3 < 10) {
                        console.print("HE CONNECTS!\n");
                        if (session.getPoints(session.getPlayer()) <= 35) session.addPlayerPoints(15);
                    } else {
                        console.print("HE MISSES \n");
                        if (session.getPoints(session.getPlayer()) != 1) console.print("\n\n");
                    }
                }
            }
            case HOOK -> {
                console.print("\n%s GIVES THE HOOK... ", session.getPlayer().getName());
                long h1 = Basic.randomOf(2);
                if (session.getOpponent().getBestPunch() == Punch.HOOK) {

                    session.addPlayerPoints(7);
                } else if (h1 == 1) {
                    console.print("BUT IT'S BLOCKED!!!!!!!!!!!!!\n");
                } else {
                    console.print("CONNECTS...\n");

                    session.addPlayerPoints(7);
                }
            }
            case UPPERCUT -> {
                console.print("\n%s  TRIES AN UPPERCUT ", session.getPlayer().getName());
                long d5 = Basic.randomOf(100);
                if (session.getOpponent().getBestPunch() == Punch.UPPERCUT || d5 < 51) {
                    console.print("AND HE CONNECTS!\n");

                    session.addPlayerPoints(4);
                } else {
                    console.print("AND IT'S BLOCKED (LUCKY BLOCK!)\n");
                }
            }
            default -> {
                console.print("%s JABS AT %s'S HEAD \n", session.getPlayer().getName(), session.getOpponent().getName());
                if (session.getOpponent().getBestPunch() == Punch.JAB) {

                    session.addPlayerPoints(3);
                } else {
                    long c = Basic.randomOf(8);
                    if (c < 4) {
                        console.print("IT'S BLOCKED.\n");
                    } else {

                        session.addPlayerPoints(3);
                    }
                }
            }
        }
    }

    private void showRoundWinner(int roundNro) {
        if (session.isRoundWinner(session.getPlayer())) {
            console.print("\n %s WINS ROUND %d\n", session.getPlayer().getName(), roundNro);
            session.addRoundWind(session.getPlayer());
        } else {
            console.print("\n %s WINS ROUND %d\n", session.getOpponent().getName(), roundNro);
            session.addRoundWind(session.getOpponent());
        }
    }

    private void showWinner() {
        if (session.isGameWinner(session.getOpponent())) {
            console.print("%s WINS (NICE GOING, " + session.getOpponent().getName() + ").", session.getOpponent().getName());
        } else if (session.isGameWinner(session.getPlayer())) {
            console.print("%s AMAZINGLY WINS!!", session.getPlayer().getName());
        } else if (session.isPlayerKnocked()) {
            console.print("%s IS KNOCKED COLD AND %s IS THE WINNER AND CHAMP!", session.getPlayer().getName(), session.getOpponent().getName());
        } else {
            console.print("%s IS KNOCKED COLD AND %s IS THE WINNER AND CHAMP!", session.getOpponent().getName(), session.getPlayer().getName());
        }

        console.print("\n\nAND NOW GOODBYE FROM THE OLYMPIC ARENA.\n");
    }

    private void loadPlayers() {
        console.print("WHAT IS YOUR OPPONENT'S NAME? ");
        final String opponentName = console.readLine();

        console.print("INPUT YOUR MAN'S NAME? ");
        final String playerName = console.readLine();

        console.print("DIFFERENT PUNCHES ARE: (1) FULL SWING; (2) HOOK; (3) UPPERCUT; (4) JAB.\n");
        console.print("WHAT IS YOUR MANS BEST? ");

        final int b = console.readInt();

        console.print("WHAT IS HIS VULNERABILITY? ");
        final int d = console.readInt();

        final Player player = new Player(playerName, Punch.fromCode(b), Punch.fromCode(d));
        final Player opponent = new Player(opponentName);

        session = new GameSession(player, opponent);
    }

    private void showIntro () {
        console.print("                                 BOXING\n");
        console.print("               CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n\n\n");
        console.print("BOXING OLYMPIC STYLE (3 ROUNDS -- 2 OUT OF 3 WINS)\n\n");
    }
}